File: BAEKJOON/4949.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

function solution(strList) {
  const [smallOpen, smallClose, bigOpen, bigClose] = ['(', ')', '[', ']'];

  strList.forEach((str) => {
    const bracketList = new Array();
    let prevBracket = null;
    let isBalance = true;

    for (char of str) {
      if (!isBalance) break; // 균형잡힌 문자열이 아니라면 break;

      switch (char) {
        case smallOpen:
          bracketList.push(smallOpen);
          break;
        case bigOpen:
          bracketList.push(bigOpen);
          break;
        case smallClose:
          prevBracket = bracketList.pop();
          if (!!prevBracket === false) isBalance = false; // 열려있는 괄호가 없을 때 ❌
          else if (prevBracket === bigOpen) isBalance = false; // 직전의 큰 괄호가 닫히기 전에 작은 괄호 닫기  ❌
          break;
        case bigClose:
          prevBracket = bracketList.pop();
          if (!!prevBracket === false) isBalance = false;
          else if (prevBracket === smallOpen) isBalance = false;
          break;
        default:
          break;
      }
    }
    if (bracketList.length !== 0) isBalance = false; // 열려있는 괄호가 남아있으면  ❌
    console.log(isBalance ? 'yes' : 'no');
  });
}

solution(input.splice(0, input.length - 1));

File: 2024-01-08/4949.py
while True:
    s = input()
    if s == '.': # 입력종료 문자
        break

    bracketStack = [] # 여는 괄호들을 담아줄 스택
    isBalance = True # VPS 여부

    for curr in s:
        if curr == '(' or curr == '[': # 여는 괄호가 나왔다면 스택에 쌓아준다.
            bracketStack.append(curr)
        elif curr == ')' or  curr == ']': # 닫힌 괄호가 나온 경우
            # 열린 괄호가 있다면 마지막 열린 괄호를 저장
            poppedBracket = bracketStack.pop() if len(bracketStack) > 0 else ''

            if (curr == ')' and poppedBracket != '(') or (curr == ']' and poppedBracket != '['): # 서로 짝이 맞지 않는 경우 VPS가 아니다.
                print('no')
                isBalance = False
                break
    
    if isBalance: # 남아있는 열린 괄호가 없을 때 VPS가 성립된다.
        print('yes' if not len(bracketStack) else 'no')